### PR TITLE
Throw exception for recursive action group

### PR DESCRIPTION
**Pull request by cliu123**
##  opensearch-security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Bug fix
 
 
 2. __Github Issue # or road-map entry, if available:__
#1389 

 3. __Description of changes:__
 If recursive action groups exist, exception is thrown while loading securityconfigs into memory.

 4. __Why these changes are required?__ 
Recursive action groups cause `StackOverflowError` while being loaded to memory.

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)
Recursive action groups cause `StackOverflowError` while security configs are being loaded to memory. 

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 

* UT
* Manual testing
  - Exception is thrown during cluster starting 

```
org.opensearch.OpenSearchSecurityException: Recursive actiongroup: recursive
	at org.opensearch.security.support.ConfigHelper.lambda$uploadFile$0(ConfigHelper.java:84) ~[opensearch-security-1.0.0.0.jar:1.0.0.0]
	at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
	at org.opensearch.security.support.ConfigHelper.lambda$uploadFile$1(ConfigHelper.java:82) ~[opensearch-security-1.0.0.0.jar:1.0.0.0]
	at java.util.HashMap.forEach(HashMap.java:1425) ~[?:?]
	at org.opensearch.security.support.ConfigHelper.uploadFile(ConfigHelper.java:80) ~[opensearch-security-1.0.0.0.jar:1.0.0.0]
	at org.opensearch.security.support.ConfigHelper.uploadFile(ConfigHelper.java:69) ~[opensearch-security-1.0.0.0.jar:1.0.0.0]
	at org.opensearch.security.configuration.ConfigurationRepository$1.run(ConfigurationRepository.java:141) [opensearch-security-1.0.0.0.jar:1.0.0.0]
	at java.lang.Thread.run(Thread.java:832) [?:?]
```
   - Security plugin is not initialized.

``` 
 % curl -XGET https://localhost:9200/_plugins/_security/api/actiongroups?pretty -u admin:admin --insecure 
OpenSearch Security not initialized.%  
```
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java [hunk 1]
```diff
@@ -147,6 +148,9 @@ private ActionGroupResolver reloadActionGroups(SecurityDynamicConfiguration<Acti
                 if(actionGroupAsObject != null && actionGroupAsObject instanceof List) {
                     
                     for (final String perm: ((List<String>) actionGroupAsObject)) {
+                        if (perm == entry) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:151]: Not a proper way to compare `Strings` in Java.
Not a proper way to identify recursion except for a self reference trivial case.

File: src/test/java/org/opensearch/security/SecurityRolesTests.java [hunk 2]
```diff
@@ -93,6 +93,24 @@ public void testSecurityRoles() throws Exception {
...
+				.setSecurityActionGroups("action_groups_recursive.yml")
+				.setSecurityRoles("roles_recursive.yml")
+				.setSecurityRolesMapping("roles_mapping_recursive.yml")
+				.setSecurityInternalUsers("internal_users_recursive.yml"), Settings.EMPTY, true);
+		});
+
+		RestHelper rh = nonSslRestHelper();
+		rh.sendAdminCertificate = false;
+
+		HttpResponse resc = rh.executeGetRequest("_cat/nodes?pretty", encodeBasicHeader("recursive_internal_user", "nagilum"));
```
Comment (vrozov) [inline src/test/java/org/opensearch/security/SecurityRolesTests.java:110]: Not a proper test, it succeeds even without code changes in `ConfigModelV7.java`.

**Review by vrozov — CHANGES_REQUESTED** [review]
-1: it is too late to raise exception **after** invalid configuration is loaded into the security index. It is necessary to validate configuration **before** it is written to the security index.

**Comment by dhiAmzn** [comment]
This only takes care of cases where action group is self referenced in its allowed actions. It will not take care of cases where there's a circular dependency such as:

```
AG_1:
  allowed_actions:
  - "AG_2"

AG_2:
  allowed_actions:
  - "AG_1"
```

**Pull request closed without merging**